### ISSUE TITLE
encode_sasl_failure

### ISSUE THREAD

**Issue by suchatorg** [processone/ejabberd#1427]
> version of ejabberd

16.12-beta1

> operating system

debian 8.6

> install ejabberd

source

Are there error messages in the log crash.log
```
2016-12-19 22:45:42 =CRASH REPORT====
  crasher:
    initial call: ejabberd_c2s:init/1
    pid: <0.234.6>
    registered_name: []
    exception exit: {{function_clause,[
{xmpp_codec,'encode_sasl_failure_$reason',
  ['bad-nonce',<<"urn:ietf:params:xml:ns:xmpp-sasl">>,[]],[{file,"src/xmpp_codec.erl"},{line,29543}]},
{xmpp_codec,encode_sasl_failure,2,[{file,"src/xmpp_codec.erl"},{line,29529}]},
{ejabberd_c2s,send_element,2,[{file,"src/ejabberd_c2s.erl"},{line,1641}]},
{ejabberd_c2s,wait_for_sasl_response,2,[{file,"src/ejabberd_c2s.erl"},{line,852}]},
{ejabberd_c2s,decode_element,3,[{file,"src/ejabberd_c2s.erl"},{line,920}]},
{p1_fsm,handle_msg,10,[{file,"src/p1_fsm.erl"},{line,582}]},
{proc_lib,init_p_do_apply,3,[{file,"proc_lib.erl"},{line,247}]}]},
[{p1_fsm,terminate,8,[{file,"src/p1_fsm.erl"},{line,760}]},
{proc_lib,init_p_do_apply,3,[{file,"proc_lib.erl"},{line,247}]}]}
    ancestors: [<0.1163.0>,ejabberd_listeners,ejabberd_sup,<0.62.0>]
    messages: []
    links: [#Port<0.208510>]
    dictionary: [{'$internal_queue_len',0}]
    trap_exit: false
    status: running
    heap_size: 46422
    stack_size: 27
    reductions: 267637
  neighbours:
```

**Comment by zinid** [comment]
This is a bug in cyrsasl_* modules. They return some made-up values, where RFC6120 clearly [forbids](https://tools.ietf.org/html/rfc6120#section-6.5) doing so. I will try to fix this to 16.12.

**Comment by zinid** [comment]
Fixed via 666608544b558dccf59ac5c29c314ff11560041b and 5ef60bf5948acd1c8b193c5312e66d4802a1ab6e
